bark_ml/observers/graph_observer.py

```diff
@@ -0,0 +1,411 @@
+      params: A `ParameterServer` instance that enables further
+        configuration of the observer. Defaults to a new instance.
+    """
+    StateObserver.__init__(self, params)
+    self._normalize_observations = normalize_observations
+
+    # the number of features of a node in the graph
+    self.feature_len = len(GraphObserver.node_attribute_keys())
+
+    # the number of features of an edge between two nodes
+    self.edge_feature_len = 4
+
+    # the maximum number of agents that can be observed
+    self._agent_limit = \
+      params["ML"]["GraphObserver"]["AgentLimit", "", 4]
+
+    # the radius an agent can 'see' in meters
+    self._visibility_radius = \
+      params["ML"]["GraphObserver"]["VisibilityRadius", "", 50]
+
+    # whether each node has an edge pointing to itself 
+    self._add_self_loops = \
+      params["ML"]["GraphObserver"]["SelfLoops", "", True]
+
+  def Observe(self, world):
+    """see base class"""
+    agents = self._preprocess_agents(world)
+    num_agents = len(agents)
+    obs = []
+
+    # append node features
+    [obs.extend(self._extract_features(agent)) for _, agent in agents]
+
+    # fill empty spots with zeros (difference between present and max number of agents)
+    obs.extend(np.zeros(max(0, self._agent_limit - num_agents) * self.feature_len))
+
+    edge_features = np.zeros((self._agent_limit, self._agent_limit, self.edge_feature_len))
+    adjacency_matrix = np.zeros((self._agent_limit, self._agent_limit))
+
+    # add edges to all visible agents
+    for index, agent in agents:
+      nearby_agents = self._nearby_agents(
+        center_agent=agent, 
+        agents=agents, 
+        radius=self._visibility_radius)
+
+      if self._add_self_loops:
+        adjacency_matrix[index, index] = 1
+
+      for target_index, nearby_agent in nearby_agents:
+        edge_features[index, target_index, :] = self._extract_edge_features(agent, nearby_agent)
+
+        adjacency_matrix[index, target_index] = 1
+        adjacency_matrix[target_index, index] = 1
+
+    adjacency_list = adjacency_matrix.reshape(-1)
+    edge_features = edge_features.reshape(-1)
+    obs.extend(adjacency_list)
+    obs.extend(edge_features)
+    
+    return tf.convert_to_tensor(obs, dtype=tf.float32, name='observation')
+
+  def _extract_edge_features(self, source_agent, target_agent):
+    """
+    Encodes the relation between the given agents as the
+    features of a directed edge in the graph.
+
+    Args:
+      source_agent: the agent that defines the source of the edge.
+      target_agent: the agent that defines the target of the edge.
+
+    Returns:
+      An np.array containing the differences in the agents' 
+      x and y position, velocities and orientations.
+    """
+    source_features = self._extract_features(source_agent, with_keys=True)
+    target_features = self._extract_features(target_agent, with_keys=True)
+
+    d_x = source_features["x"] - target_features["x"]
```

Comment:
not normed right now: could be -1 - 1 at max or 1--1 -> should be divided by 2

Reply:
if norming is wanted !!!

Reply:
definitely should enable that